### ISSUE TITLE
Embedded JS highlighting is broken

### ISSUE THREAD

**Issue by davewasmer** [atom/language-coffee-script#50]
If you use backticks (`) to escape raw JavaScript in your Coffeescript file, it seems to break the syntax highlighting / parsing.

The following line causes this:

```coffee
`import Foo from 'bar'`
```

It seems that the interior of the backticks is parsed as JavaScript (cool!), but since ES6 supports backticks for string templating, it sees the the ending backtick as the start of an ES6 template string, rather than the terminating escape character in Coffeescript (not cool).

Interestingly, this only seems to appear in Atom v0.189.0. The prior version (v0.188.0) doesn't show this problem. Not sure if this issue belongs here or in the Atom repo, let me know.

**Comment by aruseni** [comment]
I can reproduce this on my machine, too.

atom 0.189.0
language-coffee-script 0.39.0

**Comment by danii1** [comment]
Broken for me too after upgrade to atom 0.189.0

**Comment by nruth** [comment]
possible work-around: add semicolon before the closing `

**Comment by blimmer** [comment]
The semicolon is a workaround, but it's not necessary in CoffeeScript land. Can confirm that this is broken on my machine, too.

**Comment by davewasmer** [comment]
Another workaround that doesn't clutter your code is to edit the package locally. Go to the `language-coffee-script` page in your package settings, and click `View Code`:

![image](https://cloud.githubusercontent.com/assets/707213/7026715/61ac3b98-dd19-11e4-9ed5-a3563167d0b4.png)

Then, open `grammars/coffeescript.cson`, and remove the following from lines 79-81:

```
      {
        'include': 'source.js'
      }
```

You'll lose the JS highlighting of escaped JS inside your Coffeescript files, and if you update the package it will likely undo this.

**Comment by blimmer** [comment]
@davewasmer thank you for this workaround - I was about to go back to sublime this morning!

**Comment by sandstrom** [comment]
FWIW I'm seeing this too, broke in recent update of Atom.

**Comment by lnmunhoz** [comment]
This is my lines 78-81 from `grammars/coffescript.json`:
```
  },
        {
          "include": "#interpolated_coffee"
```
I can't find this cson file. Any help?

**Comment by blimmer** [comment]
I've noticed that this problem also exists for CoffeeScript PRs on Github. Must be related to an update in webkit?

**Comment by davewasmer** [comment]
My guess is that it has to do with the updates to the [language-javascript package](https://github.com/atom/language-javascript).

Atom v0.189.0 [updated](https://github.com/atom/atom/compare/v0.188.0...v0.189.0#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L136) from `language-javascript@0.64.0` to `language-javascript@0.67.0`.

Looking at [the diff between 0.64 and 0.67](https://github.com/atom/language-javascript/compare/v0.64.0...v0.67.0), it looks like lots of ES6 support was added.

Which brings us to the crux of the problem: with ES6 now using the backtick symbol for template strings, I think that means it's effectively impossible to highlight nested JS inside Coffeescript. Because if the nested JS can use the backtick symbol, it becomes impossible to tell whether the backtick is the termination of the Coffeescript escape sequence, or the start of an ES6 template string:

```
`var foo = `bar-${'fizz'}``
           ^ what does that backtick mean? It's ambiguous, even to another developer
```

I think this means that we would have to drop support for nested JS, unless there is an ES5 only version of the language grammar we could reference.

**Comment by 50Wliu** [comment]
The Github code parser uses Atom packages for some languages so it's totally possible that Github's Coffee formatting is broken since it's relying on this package.

**Comment by prenaudin** [comment]
I proposed a PR here which fixed this issue :
https://github.com/atom/language-coffee-script/pull/51

It works great when embedded JS is use (import, require for grunt, quirky things...).
Do you have comments on it ?

**Comment by xdumaine** [comment]
Ugh. The workaround works. Pain in the butt though.

**Comment by bjornstar** [comment]
Came here to complain about github's coffeescript formatting being broken for at least a week. It's driving me slowly insane.

**Comment by blimmer** [comment]
0.194.0 removed the "view code" button and I can't, for the life of me, figure out where the `coffeescript.cson` file is located on disk to reapply the fix. anyone know?

![](https://i.imgur.com/VUzYR06.png)

**Comment by davewasmer** [comment]
For me (Mac, Yosemite), it's under
`/Applications/Atom.app/Contents/Resources/app/node_modules/language-coffee-script`

**Comment by blimmer** [comment]
Are you on 0.194.0? I don't have that path (Mac, Yosemite)

**Comment by davewasmer** [comment]
Ah, no, I'm not (0.191). Perhaps it's nearby? Somewhere in the Atom.app?

**Comment by blimmer** [comment]
I've grepped the whole darn directory and can't find it.

On Wed, Apr 22, 2015 at 11:17 AM, Dave Wasmer <notifications@github.com>
wrote:

> Ah, no, I'm not (0.191). Perhaps it's nearby? Somewhere in the Atom.app?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/language-coffee-script/issues/50#issuecomment-95269940>
> .
>



-- 
*Ben Limmer*
Software Developer
ben@ibotta.com

**Comment by lnmunhoz** [comment]
Same problem as @blimmer

**Comment by mnquintana** [comment]
@blimmer That button was removed because Atom's source is now stored in a binary archive file format (.asar) that can't be inspected like it could before when it was just a directory. See this commit message for more info: https://github.com/atom/settings-view/commit/3a16b5987d1164f54e158e95111b1cd0f4de9910

**Comment by blimmer** [comment]
@mnquintana thanks for the explanation, that makes sense. 

Because of this I think we should make the change described above (or merge #51) and do a release to get compiled into the next build of atom. This is a show-stopper for use of atom + CoffeeScript in my book.

**Comment by xdumaine** [comment]
Current workaround for me:

- `git clone https://github.com/prenaudin/language-coffee-script.git`
- `cd language-coffee-script`
- `apm link`
- Restart atom

**Comment by blimmer** [comment]
Thank you @xdumaine , this workaround fixed it for me on 0.194.0 .

**Comment by mnquintana** [comment]
@xdumaine You actually don't need to restart - just reload the window (`Window: Reload` in the command palette)

**Comment by lnmunhoz** [comment]
Thank you @xdumaine

**Comment by bayri** [comment]
@xdumaine  It worked for me too. thanks!! life saver.

**Comment by sandstrom** [comment]
@xdumaine Nice workaround, thanks! :boat:

**Comment by 50Wliu** [comment]
This will be fixed in the next Atom version, 0.195.0, thanks to @prenaudin's PR.
